### PR TITLE
Add buspirate hiz and pullups feature

### PR DESCRIPTION
**Pull request by therealdreg**
This pull request introduces new functionality to the project by adding support for the Bus Pirate "HiZ mode" and "pull-up resistors" feature. 

These enhancements aim to provide users with more flexibility and control over their hardware interactions, making the tool even more versatile.

- **pullups**: Enable the Bus Pirate's built-in pull-up resistors. These resistors are useful when working **with different voltage levels**. 
- **hiz**: Enable the Bus Pirate's HiZ mode on SPI, allowing it to work as an open-collector and interface with external pull-up circuits. 

Example: Bus Pirate v3 connected to arduino uno clone ICSP + **VPU to 5V** Arduino pin:

image

- The first command works well because VPU is connected to +5V, so it is working! (5V pull-ups)
- The second command "hiz" fails because this ICSP doesn't have external pull-ups, so it is working! (open-collector standalone)
- The third command works because SPI is working in Normal Mode 3v3 with pull-ups off, so it is working!

----

image

----

This PR doesn't break anything from before and is **100% backward compatible**.

### CODE REVIEW

**Comment by therealdreg** [comment]
Please, @mcuee @MCUdude @stefanrueger Merge this soon if possible. I need it for my students to use in my next hardware hacking bootcamp :D

You can test this feature with your bus pirate v3!

**Comment by stefanrueger** [comment]
Thanks for the contribution. I read the code and it looks good to me. We normally only merge after independent verification by maintainers who have tried the mods. @MCUdude or @mcuee? Then it may take a week as merges often happen in batches.

**Comment by mcuee** [comment]
@therealdreg 

Thanks for the hardware donation.

Sorry I did not carry out the tests last time for Issue #1616 since you figured out the issue and @MCUdude did the test as well to say there was no issue with avrdude.

I will try to carry out the tests over the weekend or earlier next week.

File: src/buspirate.c
Comment (dl8dtl) [inline src/buspirate.c:66]: May I suggest a slightly different name, e.g. `BP_FLAG_PULLUPS`?
Hint: lookup "Pups" in a German dictionary. :-)

File: src/buspirate.c
Comment (dl8dtl) [inline src/buspirate.c:67]: Likewise, `BP_FLAG_HI_Z` would immediately explain to me the purpose of the bit.

**Review by dl8dtl — APPROVED** [review]
Other than this, looks good to me.

**Review by dl8dtl — APPROVED** [review]
Looks good to me.

**Comment by mcuee** [comment]
> I will try to carry out the tests over the weekend or earlier next week.

Hmm, sorry I must have placed somewhere at home and I can not find it right now. 

@MCUdude 
Just wondering if you can give this a simple test. Thanks.

**Comment by mcuee** [comment]
@MCUdude 

Just wondering if you can give this a simple test. I am sorry that I still can not find the board.

**Comment by stefanrueger** [comment]
@mcuee @MCUdude Just checking whether you plan to test this PR

**Comment by mcuee** [comment]
> @mcuee @MCUdude Just checking whether you plan to test this PR

I'd like to carry out the test but somehow I misplaced the board and still cannot find it...

**Comment by MCUdude** [comment]
Been busy with other things lately. I'll see what I can do

**Comment by MCUdude** [comment]
Am I doing something wrong? When using `-xpullups` I'm not able to communicate with the target. However, it works when not using -xpullups. I'm using the BusPirate v3.6.

```
$ avrdude -c buspirate -p atmega16 -P /dev/cu.usbserial-A10LICD2 -Uflash:w:test_files/lorem_ipsum_256B.srec -Ueeprom:w:test_files/lorem_ipsum_256B.srec
attempting to initiate BusPirate binary mode ...
avrdude: paged flash write enabled
avrdude: AVR device initialized and ready to accept instructions
avrdude: device signature = 0x1e9403 (probably m16)
avrdude: Note: flash memory has been specified, an erase cycle will be performed.
         To disable this feature, specify the -D option.
avrdude: erasing chip

avrdude: processing -U flash:w:test_files/lorem_ipsum_256B.srec:s
avrdude: reading input file test_files/lorem_ipsum_256B.srec for flash
         with 256 bytes in 1 section within [0, 0xff]
         using 2 pages and 0 pad bytes
avrdude: writing 256 bytes flash ...
Writing | ################################################## | 100% 8.26 s 
avrdude: 256 bytes of flash written
avrdude: verifying flash memory against test_files/lorem_ipsum_256B.srec
Reading | ################################################## | 100% 8.19 s 
avrdude: 256 bytes of flash verified

avrdude: processing -U eeprom:w:test_files/lorem_ipsum_256B.srec:s
avrdude: reading input file test_files/lorem_ipsum_256B.srec for eeprom
         with 256 bytes in 1 section within [0, 0xff]
         using 64 pages and 0 pad bytes
avrdude: writing 256 bytes eeprom ...
Writing | ################################################## | 100% 24.74 s 
avrdude: 256 bytes of eeprom written
avrdude: verifying eeprom memory against test_files/lorem_ipsum_256B.srec
Reading | ################################################## | 100% 8.19 s 
avrdude: 256 bytes of eeprom verified

avrdude done.  Thank you.

$ avrdude -c buspirate -p atmega16 -P /dev/cu.usbserial-A10LICD2 -Uflash:w:test_files/lorem_ipsum_256B.srec -Ueeprom:w:test_files/lorem_ipsum_256B.srec -xpullups
attempting to initiate BusPirate binary mode ...
avrdude: paged flash write enabled
avrdude: enabling pull-ups (open-collector)
avrdude error: initialization failed, rc=-2
        the programmer ISP clock is too fast for the target
        - use -B to set lower the bit clock frequency, e.g. -B 125kHz
        - use -F to override this check

avrdude done.  Thank you.
```

**Comment by therealdreg** [comment]
> Am I doing something wrong? When using `-xpullups` I'm not able to communicate with the target. However, it works when not using -xpullups. I'm using the BusPirate v3.6.
> 
> ```
> $ avrdude -c buspirate -p atmega16 -P /dev/cu.usbserial-A10LICD2 -Uflash:w:test_files/lorem_ipsum_256B.srec -Ueeprom:w:test_files/lorem_ipsum_256B.srec
> attempting to initiate BusPirate binary mode ...
> avrdude: paged flash write enabled
> avrdude: AVR device initialized and ready to accept instructions
> avrdude: device signature = 0x1e9403 (probably m16)
> avrdude: Note: flash memory has been specified, an erase cycle will be performed.
>          To disable this feature, specify the -D option.
> avrdude: erasing chip
> 
> avrdude: processing -U flash:w:test_files/lorem_ipsum_256B.srec:s
> avrdude: reading input file test_files/lorem_ipsum_256B.srec for flash
>          with 256 bytes in 1 section within [0, 0xff]
>          using 2 pages and 0 pad bytes
> avrdude: writing 256 bytes flash ...
> Writing | ################################################## | 100% 8.26 s 
> avrdude: 256 bytes of flash written
> avrdude: verifying flash memory against test_files/lorem_ipsum_256B.srec
> Reading | ################################################## | 100% 8.19 s 
> avrdude: 256 bytes of flash verified
> 
> avrdude: processing -U eeprom:w:test_files/lorem_ipsum_256B.srec:s
> avrdude: reading input file test_files/lorem_ipsum_256B.srec for eeprom
>          with 256 bytes in 1 section within [0, 0xff]
>          using 64 pages and 0 pad bytes
> avrdude: writing 256 bytes eeprom ...
> Writing | ################################################## | 100% 24.74 s 
> avrdude: 256 bytes of eeprom written
> avrdude: verifying eeprom memory against test_files/lorem_ipsum_256B.srec
> Reading | ################################################## | 100% 8.19 s 
> avrdude: 256 bytes of eeprom verified
> 
> avrdude done.  Thank you.
> 
> $ avrdude -c buspirate -p atmega16 -P /dev/cu.usbserial-A10LICD2 -Uflash:w:test_files/lorem_ipsum_256B.srec -Ueeprom:w:test_files/lorem_ipsum_256B.srec -xpullups
> attempting to initiate BusPirate binary mode ...
> avrdude: paged flash write enabled
> avrdude: enabling pull-ups (open-collector)
> avrdude error: initialization failed, rc=-2
>         the programmer ISP clock is too fast for the target
>         - use -B to set lower the bit clock frequency, e.g. -B 125kHz
>         - use -F to override this check
> 
> avrdude done.  Thank you.
> ```

@MCUdude when you use **xpullups** bus pirate works as open-collector + internal pull-ups, So:

First, you need connect VPU to external voltage or +5v pin for example.

image

image

-----

When you use **hi-z** param bus pirate works as open-collector without internal pull-ups, So, **VPU MUST BE DISCONNECTED** .

image

**Comment by stefanrueger** [comment]
> When you use xhiz param bus pirate works as open-collector without internal pull-ups, So, VPU MUST BE DISCONNECTED And in the external target circuit, there must be pull-ups.

@therealdreg Maybe it's a good idea to put something to this effect into the `.texi` documentation? Also for the pullup case.

**Comment by therealdreg** [comment]
> > When you use xhiz param bus pirate works as open-collector without internal pull-ups, So, VPU MUST BE DISCONNECTED And in the external target circuit, there must be pull-ups.
> 
> @therealdreg Maybe it's a good idea to put something to this effect into the `.texi` documentation? Both pullup and hiz cases.

done!

image

**Comment by stefanrueger** [comment]
Thanks @therealdreg Merged

**Pull request merged**
